src/impl_build_cmd.rs

```diff
@@ -0,0 +1,267 @@
+use base64::{engine::general_purpose, Engine as _};
+
+use crate::RequestFlags;
+
+const MAX_KEY_LEN: usize = 250;
+const MAX_BIN_KEY_LEN: usize = 187; // 250 * 3 / 4 due to b64 encoding
+
+// type KeyHasher = Blake2b<U18>;
+
+pub fn impl_build_cmd(
+    cmd: &[u8],
+    key: &[u8],
+    size: Option<u32>,
+    request_flags: Option<&RequestFlags>,
+    legacy_size_format: bool,
+) -> Option<Vec<u8>> {
+    if key.len() >= MAX_KEY_LEN {
+        // Key is too long
+        return None;
+    }
+    let mut binary = false;
+    for c in key.iter() {
+        if *c <= b' ' || *c > b'~' {
+            // Not ascii or containing spaces
+            binary = true;
+            break;
+        }
+    }
+    if binary && key.len() >= MAX_BIN_KEY_LEN {
+        // Key is too long
+        return None;
+    }
+
+    // Build the command
+    let mut buf: Vec<u8> = Vec::new();
+
+    // Add CMD
+    buf.extend_from_slice(cmd);
+    buf.push(b' ');
+
+    // Add key
+    if binary {
+        // If the key contains binary or spaces, it will be send in b64
+        let result = general_purpose::STANDARD.encode(key);
+        buf.extend_from_slice(&result.as_bytes());
+    } else {
+        // Otherwise, it will be send as is
+        buf.extend_from_slice(key);
+    }
+
+    // Add size
+    if let Some(size) = size {
+        buf.push(b' ');
+        if legacy_size_format {
+            buf.push(b'S');
+        }
+        buf.extend_from_slice(&size.to_string().as_bytes());
```

Comment:
If you're using size only for this part wouldn't be better to make it string instead of u32 and save the cast to string?

Reply:
Not following... We receive size u32 1234 and we need to serialize as b"1234" on the wire. You mean making str(1234) in python before passing it to the rust extension?

Reply:
@ethervoid often what happens with Rust is you end up having to follow the compiler around with casts and such. It's nice because it doesn't let you reuse buffers in risky ways.

Reply:
This is a potential crash, we should validate `size` is below some maximum if it's coming from the wire

Reply:
Actually re the crash, a 4gb (max value of a u32) buffer shouldn't crash a 64-bit process. But it's kind of excessive to send over the wire to memcache :)

Reply:
Memcache will return client error for stuff too large and the set will fail.